text/0225-configuration.md

```diff
@@ -0,0 +1,335 @@
+
+### Parse(file) -> config
+
+An API called `Parse` receives a file object. The method loads the contents of the file, parses it, and validates that the configuration against the schema. At least one of JSON or YAML MUST be supported. If either format can be supported without additional dependencies, that format SHOULD be preferred. If neither or both formats are supported, YAML should be the preferred choice. If YAML is not supported due to dependency concerns, there MAY be a way for a user to explicitly enable it by installing their own dependency.
+
+The method returns a `Configuration` model that has been validated. This API *MAY* return an error or raise an exception, whichever is idiomatic to the implementation for the following reasons:
+
+* file doesn't exist or is invalid
+* configuration parsed is invalid according to schema
+
+#### Python Parse example
+
+```python
+
+filepath = "./config.yaml"
+
+
+try:
+  cfg = opentelemetry.Parse(filepath)
+except Exception as e:
+  print(e)
+
+filepath = "./config.json"
+
+try:
+  cfg = opentelemetry.Parse(filepath)
+except Exception as e:
+  raise e
+
+```
+
+#### Go Parse example
+
+```go
+
+filepath := "./config.yaml"
+cfg, err := otel.Parse(filepath)
+if err != nil {
+  return err
+}
+
+filepath := "./config.json"
+cfg, err := otel.Parse(filepath)
+if err != nil {
+  return err
+}
+
+```
+
+Implementations *MUST* allow users to specify an environment variable to set the configuration file. This gives flexibility to end users of implementations that do not support command line arguments. The proposed name for this variable:
+
+* `OTEL_CONFIG_FILE`
+
+The format for the configuration file will be detected using the file extension of this variable.
+
+### Configuration model
+
+To allow SDKs and instrumentation libraries to accept configuration without having to implement the parsing logic, a `Configuration` model *MAY* be provided by implementations. This object:
+
+* has already been parsed from a file or data structure
+* is structurally valid (errors may yet occur when SDK or instrumentation interprets the object)
+
+### Configuration file
+
+The following demonstrates an example of a configuration file format (full example [here](https://github.com/MrAlias/otel-schema/blob/main/config.yaml)):
+
+```yaml
+# include version specification in configuration files to help with parsing and schema evolution.
+file_format: 0.1
+sdk:
+  # Disable the SDK for all signals.
+  #
+  # Boolean value. If "true", a no-op SDK implementation will be used for all telemetry
+  # signals. Any other value or absence of the variable will have no effect and the SDK
+  # will remain enabled. This setting has no effect on propagators configured through
+  # the OTEL_PROPAGATORS variable.
+  #
+  # Environment variable: OTEL_SDK_DISABLED
+  disabled: false
```

Comment:
What is the reasoning behind choosing `disabled` instead of `enabled`? 

I believe it makes sense to use the same rule across OTel in general, including the collector, which uses `enabled` not `disabled`. Will we migrate all the Collector configs from enabled?

Personally, I'd lean toward `enabled` just to avoid the double negation.

Reply:
The goal here was to align with this phrasing from the [sdk environment variable spec](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/sdk-environment-variables.md#L23-L24).

> All Boolean environment variables SHOULD be named and defined such that false is the expected safe default behavior. Renaming or changing the default value MUST NOT happen without a major version upgrade.

While this is not an environment variable, its still a reasonable position and I'd rather not rehash the lengthy debate from [#2755](https://github.com/open-telemetry/opentelemetry-specification/pull/2755). 🙂

Reply:
Do we have to apply to same rule for booleans in the yaml configs, given that we have this note?

> Note that there is no consistent mapping between environment variable names and the keys in the configuration file.

Reply:
> Do we have to apply to same rule for booleans in the yaml configs, given that we have this note?

We don't have to, but IMO we should.

But regardless of what my opinion is on this subject, we'd like to keep the exact schema of the configuration file is out of scope for this OTEP. This merely illustrates an example. The exact details will have to be hashed out later in the specification. @MrAlias is [going to add](https://github.com/open-telemetry/oteps/pull/225#discussion_r1142339164) text saying that the exact schema of the config file is an open question.

Reply:
@jpkrohling Thanks for the feedback, definitely something to ponder. This OTEP doesn't take a strong position on this, and as Tyler mentioned, this is included in the Open Questions section, so we will have to discuss that once this hits the Spec.